### PR TITLE
fix: mark cascading teardown-hook failures as secondary

### LINKED ISSUE #5925: 🐛 Bug: parent-suite teardown error masks original failure from nested suite
### Bug Report Checklist

- [x] This is NOT a [security, `npm audit`, or GitHub Advisory issue](https://mochajs.org/explainers/security-vulnerability-reports).
- [x] I have read and agree to Mocha's [Code of Conduct](https://github.com/mochajs/mocha/blob/main/.github/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/mochajs/mocha/blob/main/.github/CONTRIBUTING.md)
- [x] I have searched for [related issues](https://github.com/mochajs/mocha/issues?q=is%3Aissue) and [issues with the `faq` label](https://github.com/mochajs/mocha/issues?utf8=%E2%9C%93&q=is%3Aissue%20label%3Afaq%20), but none matched my issue.
- [x] I have 'smoke tested' the code to be tested by running it outside the real test suite to get a better sense of whether the problem is in the code under test, my usage of Mocha, or Mocha itself.
- [ ] I want to provide a PR to resolve this

### Expected

When a test or hook inside a nested `describe()` fails, and a teardown hook registered on a parent suite also fails afterward, Mocha should keep the original nested failure as the primary reported error.

The parent teardown error is secondary. It should not obscure the failure that actually caused the test run to become unhealthy.

This is a common layout in larger test suites:

```js
describe('outer suite', () => {
  afterEach(() => cleanup())

  describe('inner suite', () => {
    beforeEach(() => setup())

    it('does something', () => {})
  })
})
```

If the nested setup/test fails, and the parent cleanup also throws, the original nested failure is the useful root cause.

### Actual

Mocha reports the parent teardown failure as an additional failure after the original nested failure. In large CI logs this makes the teardown cascade look like an independent failure and can obscure the actual root cause.

```
    outer suite
      inner suite
        1) "before each" hook for "does something"
      2) "after each" hook for "does something"

    0 passing
    2 failing

    1) outer suite inner suite
         "before each" hook for "does something":
       Error: inner setup failed

    2) outer suite
         "after each" hook for "does something":
       Error: outer cleanup failed
```

The second error is a teardown cascade. The first error is the root cause.

### Minimal, Complete and Verifiable Example

```js
const { describe, beforeEach, afterEach, it } = require('mocha')

describe('outer suite', () => {
  afterEach(() => {
    throw new Error('outer cleanup failed')
  })

  describe('inner suite', () => {
    beforeEach(() => {
      throw new Error('inner setup failed')
    })

    it('does something', () => {})
  })
})
```

```js
const { describe, after, it } = require('mocha')

describe('outer suite', () => {
  after(() => {
    throw new Error('outer cleanup failed')
  })

  describe('inner suite', () => {
    it('fails for the real reason', () => {
      throw new Error('inner test failed')
    })
  })
})
```

### Versions

- Mocha 11.7.5
- Node.js v24.13.0

### Additional Info

This is related to #5922, but the nested-suite case is slightly different: the original failure is not necessarily in the same suite where the teardown hook was registered.

### PR DESCRIPTION
**Pull request by naba-runn**
- [x] Addresses an existing open issue: fixes #5925
- [x] That issue was marked as [`status: accepting prs`](https://github.com/mochajs/mocha/issues?q=is%3Aopen+is%3Aissue+label%3A%22status%3A+accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/mochajs/mocha/blob/main/.github/CONTRIBUTING.md) were taken

When a test/hook fails in a nested suite and a parent teardown hook (afterEach/after all) also throws, both show up as separate failures and the teardown one hides the real cause.

This PR flags the teardown error as secondary and shows a (secondary) label in the reporter, so the original failure stays primary. Failure count stays same, and independent teardown failures are not affected.

Also added tests and fixtures.

Another clarification, I went with keeping the count same, just labeled it secondary. Another way would be to merge it into the original failure so count becomes "1 failing" but there is a risk of breaking count parsing. Let me know if you would prefer the second option! Also, if there's any other change you want in this PR, please do let me know, I am more than happy to contribute to this repo:) Thank you!

### CODE REVIEW

File: lib/reporters/base.js [hunk 1]
```diff
@@ -532,6 +532,11 @@ Base.list = function (failures) {
       testTitle += str;
     });
 
+    // Label teardown cascades so the original failure stays the primary one (#5925)
```
Comment (mark-wiemer) [inline lib/reporters/base.js]:
```suggestion
```

File: test/integration/hook-err.spec.cjs [hunk 2]
```diff
@@ -130,6 +130,63 @@ describe("hook error handling", function () {
...
+  describe("teardown error masks original failure from nested suite", function () {
+    it("marks a cascading afterEach failure as secondary", function (done) {
+      runMochaJSON(
+        "hooks/after-each-hook-secondary-error.fixture.js",
+        [],
+        (err, res) => {
+          if (err) {
+            return done(err);
+          }
+          expect(res, "to have failed test count", 2);
```
Comment (mark-wiemer) [inline test/integration/hook-err.spec.cjs:143]: I know I'm missing something: why would this report 2 failed tests when only one test is being run?

@copilot feel free to chime in

**Comment by mark-wiemer** [comment]
@naba-runn please remove the verbose comments from the PR and esp remove references to particular PRs, we can find those later in the blame if needed :)

**Comment by naba-runn** [comment]
@mark-wiemer hey so there is only one test case, but two hooks that throw errors(beforeEach and afterEach). The actual test never runs because the beforeEach aborts it before it reaches there. This is the same thing I mentioned in the description earlier, right now I am keeping the count as 2 and just labelling the teardown one as secondary. If you would prefer to show only 1 failing, I will merge it into the original error(beforeEach error), I will be happy to make the change:) Thank you!

**Comment by naba-runn** [comment]
Also removed the verbose comments and added the else { delete err.secondary; } as suggested by copilot. I will add the regression case as well.


> * The cascade marker is only set for errors delivered as `cbHookRun(err)`. Errors emitted through a hook's `error` listener (for example the multiple-done/late-async path, which calls `self.fail(hook, err)` directly at line 628) bypass this block, so a parent teardown failure after an earlier nested failure is still reported without `secondary`. Apply the same check before that direct-fail path, ideally through a shared helper.
> 
> ```
>           self._failSecondary = true;
>         }
> ```
Also I looked into this and I did miss this edge case, but as the path is asynchronous, the runner current test may have moved ahead by the time error listener fires. So we can't use the same failure count check here. Also this is mostly an edge case path, the normal throw path is already handled so it would be better to handle this in a follow up. But if you want this here, I am up for it. Thank you!

**Comment by naba-runn** [comment]
Fixed the codecov error and added the regression test that copilot suggested:)

File: lib/runner.js [hunk 3]
```diff
@@ -553,6 +559,12 @@ Runner.prototype.fail = function (test, err, force) {
     err = thrown2Error(err);
   }
 
+  if (secondary) {
+    err.secondary = true;
+  } else {
+    delete err.secondary;
+  }
```
[the remark this answers is not in the document]
Comment (mark-wiemer) [reply lib/runner.js:566]: I'm not 100% sure this is legit, I will try to repro later. Overall this is a complex PR that's not a huge priority, I will get to it eventually :)
